pkg/broker/broker_test.go

```diff
@@ -64,3 +65,176 @@ func ExampleBrokerService_IsEnabled() {
 	// Output: true
 	// false
 }
+
+func ExampleBrokerService_ServiceDefinition() {
+	service := BrokerService{
+		Name: "left-handed-smoke-sifter",
+		DefaultServiceDefinition: `{"id":"abcd-efgh-ijkl"}`,
+	}
+
+	// Default definition
+	defn, err := service.ServiceDefinition()
+	fmt.Printf("%q %v\n", defn.ID, err)
+
+	// Override
+	viper.Set(service.DefinitionProperty(), `{"id":"override-id"}`)
+	defn, err = service.ServiceDefinition()
+	fmt.Printf("%q %v\n", defn.ID, err)
+
+	// Bad Value
+	viper.Set(service.DefinitionProperty(), "nil")
+	_, err = service.ServiceDefinition()
+	fmt.Printf("%v\n", err == nil)
+
+	// Cleanup
+	viper.Set(service.DefinitionProperty(), nil)
+
+	// Output: "abcd-efgh-ijkl" <nil>
+	// "override-id" <nil>
+	// false
+}
+
+func ExampleBrokerService_GetPlanById() {
+	service := BrokerService{
+		Name: "left-handed-smoke-sifter",
+		DefaultServiceDefinition: `{"id":"abcd-efgh-ijkl", "plans": [{"id": "builtin-plan", "name": "Builtin!"}]}`,
+	}
+
+	viper.Set(service.UserDefinedPlansProperty(), `[{"id":"custom-plan", "name": "Custom!"}]`)
+	defer viper.Set(service.UserDefinedPlansProperty(), nil)
+
+	plan, err := service.GetPlanById("builtin-plan")
+	fmt.Printf("%q %v\n", plan.Name, err)
+
+	plan, err = service.GetPlanById("custom-plan")
+	fmt.Printf("%q %v\n", plan.Name, err)
+
+	_, err = service.GetPlanById("missing-plan")
+	fmt.Printf("%s\n", err)
+
+	// Output: "Builtin!" <nil>
+	// "Custom!" <nil>
+	// Plan ID "missing-plan" could not be found
+}
+
+func TestBrokerService_UserDefinedPlans(t *testing.T) {
+	cases := map[string]struct {
+		Value       interface{}
+		PlanCount   int
+		ExpectError bool
+	}{
+		"default-no-plans": {
+			Value:       nil,
+			PlanCount:   0,
+			ExpectError: false,
+		},
+		"single-plan": {
+			Value:       `[{"id":"aaa"}]`,
+			PlanCount:   1,
+			ExpectError: false,
+		},
+		"bad-json": {
+			Value:       `42`,
+			PlanCount:   0,
+			ExpectError: true,
+		},
+		"multiple-plans": {
+			Value:       `[{"id":"aaa"},{"id":"bbb"}]`,
+			PlanCount:   2,
+			ExpectError: false,
+		},
+	}
+
+	service := BrokerService{
+		Name: "left-handed-smoke-sifter",
+		DefaultServiceDefinition: `{"id":"abcd-efgh-ijkl"}`,
+	}
+
+	for tn, tc := range cases {
+		viper.Set(service.UserDefinedPlansProperty(), tc.Value)
+		plans, err := service.UserDefinedPlans()
+
+		// Check errors
+		hasErr := err != nil
+		if hasErr != tc.ExpectError {
+			t.Errorf("%s) Expected Error? %v, got error: %v", tn, tc.ExpectError, err)
+		}
+
+		// Check IDs
+		if len(plans) != tc.PlanCount {
```

Comment:
In addition to verifying the correct number of plans, it could be helpful to also verify the plan IDs were correctly propagated. Perhaps adding expectedIDs to the test case struct.

Reply:
Done, I really miss real set arithmetic. One of the few places I really think the go authors made a mistake.